### PR TITLE
libvirtd: added QEMU hook option

### PR DESCRIPTION
**Pull request by anpin**
## Description of changes

Added option and configuration handler for [libvirtd QEMU hooks](https://www.libvirt.org/hooks.html) 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/virtualisation/libvirtd.nix [hunk 1]
```diff
@@ -127,6 +127,14 @@ let
           QEMU's swtpm options.
         '';
       };
+
+      hook = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = lib.mdDoc ''
+          QEMU's hook script.
```
Comment (tmplt) [inline nixos/modules/virtualisation/libvirtd.nix:135]: I think the documentation should cover the `PATH` gotchas mentioned [here](https://discourse.nixos.org/t/libvirt-installing-qemu-hook/385/8?u=tmplt).

**Review by tmplt — COMMENTED** [review]
Thanks for upstreaming this.

File: nixos/modules/virtualisation/libvirtd.nix [hunk 2]
```diff
@@ -337,6 +345,17 @@ in
           ln -s --force ${ovmfpackage}/FV/AAVMF_VARS.fd /run/${dirName}/nix-ovmf/
           ln -s --force ${ovmfpackage}/FV/OVMF_VARS.fd /run/${dirName}/nix-ovmf/
         '')}
+
+        ${if cfg.qemu.hook != null then ''
+          # copy qemu hook script
+          mkdir -p /var/lib/libvirt/hooks
+          chmod 755 /var/lib/libvirt/hooks
+          cp -f ${cfg.qemu.hook} /var/lib/libvirt/hooks/qemu
```
Comment (fpletz) [inline nixos/modules/virtualisation/libvirtd.nix:353]:
```suggestion
          ln -sf ${cfg.qemu.hook} /var/lib/libvirt/hooks/qemu
```

File: nixos/modules/virtualisation/libvirtd.nix [hunk 3]
```diff
@@ -337,6 +345,17 @@ in
           ln -s --force ${ovmfpackage}/FV/AAVMF_VARS.fd /run/${dirName}/nix-ovmf/
           ln -s --force ${ovmfpackage}/FV/OVMF_VARS.fd /run/${dirName}/nix-ovmf/
         '')}
+
+        ${if cfg.qemu.hook != null then ''
+          # copy qemu hook script
+          mkdir -p /var/lib/libvirt/hooks
+          chmod 755 /var/lib/libvirt/hooks
```
Comment (fpletz) [inline nixos/modules/virtualisation/libvirtd.nix:352]:
```suggestion
          mkdir -p -m 755 /var/lib/libvirt/hooks
```

File: nixos/modules/virtualisation/libvirtd.nix (hunk 1 above)
Comment (fpletz) [reply nixos/modules/virtualisation/libvirtd.nix:135]: We can mention this but people using NixOS should know this. :sweat_smile:

**Review by fpletz — CHANGES_REQUESTED** [review]

File: nixos/modules/virtualisation/libvirtd.nix (hunk 1 above)
Comment (tmplt) [reply nixos/modules/virtualisation/libvirtd.nix:135]: I don't consider "people should know this" a valid reason to omit useful documentation on details out of view from the user.

**Comment by jian-lin** [comment]
FYI, a previous [PR](https://github.com/NixOS/nixpkgs/pull/232250), which is covers more kinds of hooks.

cc @YBeaugnon

**Comment by fpletz** [comment]
Also didn't notice that PR before. Thanks everyone.

**Pull request closed without merging**